### PR TITLE
CP-54476 Merge ntp disable/enable to ntp mode

### PR DESCRIPTION
**Pull request by changlei-li**
As the doc PR https://github.com/xapi-project/xen-api/pull/6745 shows, the enable/disable ntp APIs and host.ntp_enabled field can be merged into host.ntp_mode.
Now, the ntp modes are DHCP, Custom, Factory and Disabled.
This PR also updates the datamodel lifecycle and bumps up the schema.

### CODE REVIEW

**Comment by changlei-li** [comment]
Test:
```python
    # set NTP mode Custom, use custom NTP servers
    session.xenapi.host.set_ntp_custom_servers(host_ref, ['time.google.com', 'time.windows.com', 'time.apple.com'])
    session.xenapi.host.set_ntp_mode(host_ref, 'Custom')
    time.sleep(5)
    show('NTP mode Custom, use custom NTP servers')

    # set NTP mode DHCP
    session.xenapi.host.set_ntp_mode(host_ref, 'DHCP')
    time.sleep(5)
    show('set NTP mode DHCP')

    # set NTP mode Factory
    session.xenapi.host.set_ntp_mode(host_ref, 'Factory')
    time.sleep(5)
    show('set NTP mode Factory')

    # disable NTP
    session.xenapi.host.set_ntp_mode(host_ref, 'Disabled')
    session.xenapi.host.set_servertime(host_ref, '20251125T14:11:55+08:00')
    show('set NTP mode Disbaled and set server time')

    # enable NTP and set to DHCP mode
    session.xenapi.host.set_ntp_mode(host_ref, 'DHCP')
    time.sleep(5)
    show('set NTP mode DHCP')
```
output:
```
===NTP mode Custom, use custom NTP servers===
ntp_mode: Custom; timezone: UTC; ntp_synchronized: True; servertime: 20251125T07:20:12Z
ntp_custom_servers: ['time.google.com', 'time.windows.com', 'time.apple.com']
ntp_server_status: {'216.239.35.0': 'synced', '168.61.215.74': 'uncombined', '17.253.2.35': 'uncombined'}
===set NTP mode DHCP===
ntp_mode: DHCP; timezone: UTC; ntp_synchronized: True; servertime: 20251125T07:20:19Z
ntp_custom_servers: ['time.google.com', 'time.windows.com', 'time.apple.com']
ntp_server_status: {'10.63.96.11': 'synced'}
===set NTP mode Factory===
ntp_mode: Factory; timezone: UTC; ntp_synchronized: True; servertime: 20251125T07:20:26Z
ntp_custom_servers: ['time.google.com', 'time.windows.com', 'time.apple.com']
ntp_server_status: {'134.215.155.177': 'combined', '216.229.4.69': 'synced', '69.176.84.38': 'combined', '144.34.193.110': 'combined'}
===set NTP mode Disbaled and set server time===
ntp_mode: Disabled; timezone: UTC; ntp_synchronized: False; servertime: 20251125T06:11:56Z
ntp_custom_servers: ['time.google.com', 'time.windows.com', 'time.apple.com']
ntp_server_status: {}
===set NTP mode DHCP===
ntp_mode: DHCP; timezone: UTC; ntp_synchronized: True; servertime: 20251125T07:20:35Z
ntp_custom_servers: ['time.google.com', 'time.windows.com', 'time.apple.com']
ntp_server_status: {'10.63.96.11': 'synced'}
```

Exceptions example:
```python
    try:
        session.xenapi.host.set_ntp_mode(host_ref, 'Custom')
        session.xenapi.host.set_ntp_custom_servers(host_ref, [])
    except Exception as e:
        print(e)
    try:
        session.xenapi.host.set_ntp_mode(host_ref, 'DHCP')
        session.xenapi.host.set_ntp_custom_servers(host_ref, [])
        session.xenapi.host.set_ntp_mode(host_ref, 'Custom')
    except Exception as e:
        print(e)
    try:
        session.xenapi.host.set_servertime(host_ref, '20251125T14:11:55+08:00')
    except Exception as e:
        print(e)
```
output
```
['INVALID_NTP_CONFIG', "Can't set ntp_custom_servers empty when ntp_mode is Custom"]
['INVALID_NTP_CONFIG', "Can't set ntp_mode to Custom when ntp_custom_servers is empty"]
['NOT_ALLOWED_WHEN_NTP_IS_ENABLED', 'OpaqueRef:91f90f95-3a9b-d048-6f92-a4e522654540']
```

File: ocaml/xapi/xapi_host.ml [hunk 1]
```diff
@@ -3471,90 +3471,90 @@ let sync_max_cstate ~__context ~host =
 
 let set_ntp_mode ~__context ~self ~value =
   let current_mode = Db.Host.get_ntp_mode ~__context ~self in
-  let ntp_enabled = Db.Host.get_ntp_enabled ~__context ~self in
   if current_mode <> value then (
     let open Xapi_host_ntp in
-    let ensure_custom_servers_exist servers =
+    let ensure_servers_exist servers msg =
```
Comment (minglumlu) [inline ocaml/xapi/xapi_host.ml:3476]: I think at least 3 NTP servers are required.
And it could be `assert_sufficient_ntp_servers`.

File: ocaml/xapi/xapi_host.ml [hunk 2]
```diff
@@ -3471,90 +3471,90 @@ let sync_max_cstate ~__context ~host =
...
+    in
+    let enter_mode = function
+      | `DHCP ->
+          add_dhcp_ntp_servers ()
+      | `Custom ->
+          let custom_servers =
+            Db.Host.get_ntp_custom_servers ~__context ~self
+          in
+          ensure_servers_exist custom_servers
+            "Can't set ntp_mode to Custom when ntp_custom_servers is empty" ;
```
Comment (minglumlu) [inline ocaml/xapi/xapi_host.ml:3488]: The two messages `... when ntp_custom_servers is empty` and `... when factory ntp servers are empty` should be consistent.

File: ocaml/xapi/xapi_host.ml (hunk 1 above)
Comment (changlei-li) [reply ocaml/xapi/xapi_host.ml:3476]: Ah, I find rfc8633 in chapter `3.2. Using Enough Time Sources` recommends the four or more time sources. But here, I think it's just the best practice, not a hard requirement.

File: ocaml/xapi/xapi_host.ml [hunk 3]
```diff
@@ -3471,90 +3471,90 @@ let sync_max_cstate ~__context ~host =
...
-    | _, _ ->
-        ()
+    | `Disabled, next ->
+        enter_mode next ; enable_ntp_service ()
+    | _, `Disabled ->
+        exit_mode current_mode ; disable_ntp_service ()
+    | current, next ->
+        exit_mode current ;
+        enter_mode next ;
+        Xapi_host_ntp.restart_ntp_service ()
```
Comment (minglumlu) [inline ocaml/xapi/xapi_host.ml:3514]: Would it be safer to call `enable_ntp_service` before this restart to always get expected result? For example for some reason the ntp is set to `false`.

**Review by minglumlu — APPROVED** [review]

File: ocaml/xapi/xapi_host.ml (hunk 1 above)
Comment (minglumlu) [reply ocaml/xapi/xapi_host.ml:3476]: 1 or 2 will fail definitely. This is for how NTP works.

File: ocaml/xapi/xapi_host.ml (hunk 2 above)
Comment (changlei-li) [reply ocaml/xapi/xapi_host.ml:3488]: You mean add two api errors or just move the strings to xapi_globs?
Now I add one api error `invalid_ntp_config` in api_errors, the string is reason.
```
  error Api_errors.invalid_ntp_config ["reason"]
    ~doc:"The NTP configuration is invalid." () ;
```

File: ocaml/xapi/xapi_host.ml (hunk 3 above)
Comment (changlei-li) [reply ocaml/xapi/xapi_host.ml:3514]: It's OK. In fact, the `exit_mode` and `enter_mode` operations are manipulating chronyd conf. The operations can be done even NTP is disabled.
For example:
```
[root@nkg-dt15 ~]# xe host-param-get uuid=3c9c8fe0-2175-4e98-9ee8-f27dd6e231f3 param-name=ntp-mode
DHCP
[root@nkg-dt15 ~]# timedatectl set-ntp false
[root@nkg-dt15 ~]# xe host-param-set uuid=3c9c8fe0-2175-4e98-9ee8-f27dd6e231f3 ntp-mode=Factory
[root@nkg-dt15 ~]# chronyc sources
MS Name/IP address         Stratum Poll Reach LastRx Last sample               
===============================================================================
^+ ntp2.ggsrv.de                 2   6    17    12  -5169us[-5810us] +/-  128ms
^* 157.245.102.2                 2   6    17    12   -285us[ -926us] +/-   56ms
^- ntp7.mum-in.hosts.301-mo>     2   6    17    12  +2176us[+2176us] +/-  121ms
^+ time.cloudflare.com           3   6    17    14  +3026us[+2385us] +/-   81ms
```

File: ocaml/xapi/xapi_host.ml (hunk 1 above)
Comment (changlei-li) [reply ocaml/xapi/xapi_host.ml:3476]: I think "1 or 2 will fail definitely" is too strong. They are risky and uncertain but not guaranteed failure.
One server: If it goes down or drifts significantly, there’s no backup.
Two servers: If they disagree, NTP can’t determine which one is correct.

File: ocaml/xapi/xapi_host.ml (hunk 2 above)
Comment (minglumlu) [reply ocaml/xapi/xapi_host.ml:3488]: No. Only for the wording. like `ntp_custom_servers` vs `factory ntp servers`, and `is` vs `are`.

File: ocaml/xapi/xapi_host.ml (hunk 3 above)
Comment (minglumlu) [reply ocaml/xapi/xapi_host.ml:3514]: Will the NTP be set to true when NTP is false and after calling restart_ntp_service ()?

Comment (changlei-li) [reply ocaml/xapi/xapi_host.ml:3514]: For timedatectl NTP flag, yes. But maybe need a few minutes, the same reason as the second commit, see the second commit message.

File: ocaml/xapi/xapi_host.ml (hunk 2 above)
Comment (changlei-li) [reply ocaml/xapi/xapi_host.ml:3488]: Regarding servers as list, both use `is`.

File: ocaml/xapi/xapi_host.ml (hunk 3 above)
Comment (changlei-li) [reply ocaml/xapi/xapi_host.ml:3514]: Sorry. I have a test, the `NTP enabled` is still `no` in `timedatectl status`. The `NTP synchronized` is detected periodically but not NTP enabled. This may affect the set-ntp-mode Disabled in some corner case. I will change to use your suggestion.

File: ocaml/xapi/xapi_host.ml (hunk 1 above)
Comment (psafont) [reply ocaml/xapi/xapi_host.ml:3476]: From the description of what happens with one or two servers, it looks to me NTP is unable to work reliably :)

Because we want xapi to work reliably at all times, I think it's warranted to enforce at least 3 servers.

File: ocaml/xapi/xapi_host.ml (hunk 3 above)
Comment (changlei-li) [reply ocaml/xapi/xapi_host.ml:3514]: Done. Fix in function Xapi_host_ntp.restart_ntp_service ()

File: ocaml/xapi/xapi_host.ml (hunk 1 above)
Comment (changlei-li) [reply ocaml/xapi/xapi_host.ml:3476]: I still insist we shouldn't strict the API settings. NTP service like chornyd will not limit the number of time sources. It's strange to encounter this error in our APIs if I am the user. If in user's internal network environment, there is only one NTP server deployed, it is hard to work.
@robhoes Will I get a vote from you on this issue.

File: ocaml/idl/datamodel_host.ml [hunk 4]
```diff
@@ -3297,15 +3284,12 @@ let t =
             ~default_value:(Some (VBool false)) "secure_boot"
             "Whether the host has booted in secure boot mode"
         ; field ~qualifier:DynamicRO ~lifecycle:[] ~ty:host_ntp_mode
-            ~default_value:(Some (VEnum "ntp_mode_dhcp")) "ntp_mode"
+            ~default_value:(Some (VEnum "DHCP")) "ntp_mode"
```
Comment (robhoes) [inline ocaml/idl/datamodel_host.ml]: The default is set to DHCP here, which is relevant for the update case. This implies that before this feature, hosts relied on DHCP to set up NTP. Is that right, as I expected `Factory`?

File: ocaml/xapi/xapi_host.ml (hunk 1 above)
Comment (robhoes) [reply ocaml/xapi/xapi_host.ml:3476]: I would prefer to not restrict this in xapi.

File: ocaml/xapi/xapi_host.ml [hunk 5]
```diff
@@ -3471,90 +3471,90 @@ let sync_max_cstate ~__context ~host =
...
+          in
+          ensure_servers_exist custom_servers
+            "Can't set ntp_mode to Custom when ntp_custom_servers is empty" ;
+          set_servers_in_conf custom_servers
+      | `Factory ->
+          let factory_servers = !Xapi_globs.factory_ntp_servers in
+          ensure_servers_exist factory_servers
+            "Can't set ntp_mode to Factory when factory ntp servers is empty" ;
+          set_servers_in_conf factory_servers
+      | `Disabled ->
```
Comment (minglumlu) [inline ocaml/xapi/xapi_host.ml:3495]: I forgot one thing. Should the system clock be synced to RTC hardware clock when NTP is about to be disabled?

File: ocaml/idl/datamodel_host.ml (hunk 4 above)
Comment (minglumlu) [reply ocaml/idl/datamodel_host.ml]: It seems we have to rely on the `Dbsync_slave.update_env` to get the info when xapi restart for the update application.

File: ocaml/xapi/xapi_host.ml (hunk 5 above)
Comment (changlei-li) [reply ocaml/xapi/xapi_host.ml:3495]:
```
# Enable kernel synchronization of the real-time clock (RTC).
rtcsync
```
In our dom0 host, chrony is configured to sync RTC after NTP synchronized. So there is no need to do it again in XAPI. I also have verified it.

File: ocaml/idl/datamodel_host.ml (hunk 4 above)
Comment (changlei-li) [reply ocaml/idl/datamodel_host.ml]: Done. Default value of ntp_mode changes to Factory.

File: ocaml/xapi/xapi_host.ml (hunk 1 above)
Comment (psafont) [reply ocaml/xapi/xapi_host.ml:3476]: Then I'd prefer if the standard recommending at least 3 servers was added in the datamodel field. This is something I'll also mention to XO people to avoid issues
